oxidation/libparsec/crates/types/src/regex.rs

```diff
@@ -12,43 +31,131 @@ fn escape_globing_pattern(string: &str) -> String {
+    pub fn from_raw_regexes(raw_regexes: &[&str]) -> RegexResult {
+        Ok(Self(
+            raw_regexes
+                .iter()
+                .map(|l| regex::Regex::new(l).map_err(|err| RegexError::ParseError { err }))
+                .collect::<Result<Vec<regex::Regex>, RegexError>>()?,
+        ))
     }
 
-    pub fn is_match(&self, string: &str) -> bool {
-        self.0.is_match(string)
+    /// Returns a regex from a glob pattern
+    pub fn from_glob_pattern(pattern: &str) -> RegexResult {
+        let escaped_str = escape_globing_pattern(pattern);
+        Ok(Regex(vec![fnmatch_regex::glob_to_regex(&escaped_str)
+            .map_err(|err| RegexError::GlobPatternError { err })?]))
+    }
+
+    pub fn from_regex_str(regex_str: &str) -> RegexResult {
+        Self::from_raw_regexes(&[regex_str])
     }
-}
 
-impl AsRef<str> for Regex {
-    fn as_ref(&self) -> &str {
-        self.0.as_str()
+    pub fn is_match(&self, string: &str) -> bool {
+        self.0.iter().any(|r| r.is_match(string))
     }
 }
 
 impl PartialEq for Regex {
     fn eq(&self, other: &Self) -> bool {
-        self.as_ref() == other.as_ref()
+        if self.0.len() == other.0.len() {
+            let set: HashSet<&str> = HashSet::from_iter(self.0.iter().map(|r| r.as_str()));
+
+            other.0.iter().all(|r| set.contains(r.as_str()))
+        } else {
+            false
+        }
     }
 }
 
 impl Display for Regex {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        self.0.fmt(f)
+        for (i, r) in self.0.iter().enumerate() {
+            r.fmt(f)?;
+
+            if i + 1 < self.0.len() {
+                f.write_char('|')?;
+            }
+        }
+
+        Ok(())
     }
 }
 
 #[cfg(test)]
 mod tests {
+    use std::{fs, path::Path};
+
     use crate::regex::Regex;
     use rstest::rstest;
 
+    #[rstest]
+    #[case::base("*.rs\n*.py", "base.tmp")]
+    #[case::trim_whitespace("  *.rs\n   *.py   ", "trim_whitespace.tmp")]
+    #[case::empty_lines("*.rs\n\n\n*.py", "empty_lines.tmp")]
+    #[case::trim_whitespace_and_empty_lines(
+        "   *.rs\n\n  \n\n*.py  ",
+        "trim_whitespace_and_empty_lines.tmp"
+    )]
+    #[case::ignore_comment(
+        "# This contains patterns\n## yes\n   *.rs\n\n  \n\n*.py  ",
+        "ignore_comment.tmp"
+    )]
+    fn from_pattern_file_content(#[case] file_content: &str, #[case] filename: &str) {
+        fs::write(filename, file_content).unwrap();
+        let regex = Regex::from_file(Path::new(filename)).expect("Regex should be valid");
+        fs::remove_file(filename).unwrap();
+
+        assert!(regex.is_match("file.py"));
+        assert!(regex.is_match("file.rs"));
+        assert!(!regex.is_match("stories.txt"));
+    }
+
+    #[test]
+    fn load_default_pattern_file() {
+        let regex = Regex::from_file(Path::new(
+            "../../../../parsec/core/resources/default_pattern.ignore",
```

Comment:
And windows ?

Reply:
The test seems to work just fine :thinking:

Reply:
https://github.com/Scille/parsec-cloud/actions/runs/3657049957/jobs/6180500459#step:8:804